docs/lsm/allocation/allocation_sources/allocation_v3/basic_example/v3_main.cf

```diff
@@ -0,0 +1,81 @@
+"""
+    Inmanta LSM
+    :copyright: 2024 Inmanta
+    :contact: code@inmanta.com
+    :license: Inmanta EULA
+"""
+
+import lsm
+import lsm::fsm
+
+entity TopLevelService extends lsm::ServiceEntity:
+    """
+    This service entity demonstrates how a single allocator
+    can be used for both a service entity and its embedded
+    entities.
+    """
+
+    string                      name
+    lsm::attribute_modifier     name__modifier="rw"
+    int?                        top_level_value
```

Comment:
I thought we always needed to have a default, null, value for the allocated attributes, isn't it the case? Which value will the api assign when the service is created by the user otherwise?

Reply:
The API will always default to `None` for nullable attributes, if no explicit default is given. I'm not sure how I feel about that but it is the current behavior as far as I can tell.

Reply:
I partially agree and partially don't, here is some more food for thoughts:
1. I agree that "nullable" and "optional" are different concepts.  I associate "optional" to "has a default value".  Assuming any defined attribute should have a value, if you allow (or force in case of `ro` attribute) the user to not specify a value, making the attribute optional in the input, then that attribute should define a default, which you can then use when the entity is constructed (in the api).  In some cases (talking about javascript here), a defined attribute might not have a value (`undefined`), but this is not a thing that exists in our DSL, and I think we should not make it a thing in our API as it is defined in the DSL.
2. I agree that defaults complicates things in the DSL in general, but I would say it is not the case for lsm entities, as whenever we construct them, we will always have all their attributes at our disposal (as we read them from the api, or from an allocator).
3. Regarding the `nullable` aspect of allocated values, I would say it should indeed not be required, but I believe that `null` is a generic enough default for a value that is not provided (yet).  And it is actually a convention we expect users to follow when using the `allocator` decorator, as this is the "no value" we check for, before calling the decorated plugin.  In the future, we will introduce the `reallocator` decorator, which will be called in any validating compile, providing the previous value that was set in the api.  With this mechanism, the user could define another default (i.e. `"unknown"`) that the decorated plugin can recognize and replace with the real allocated value.

Reply:
3. Doesn't this ever give trouble in the model? e.g. suppose you have a service that writes files, but the file name is allocated. It would be an invariant of the model that the file name can never be `null`: either it's provided through the API (`lsm::all`) or it is allocated. Yet you can not express this invariant in the model because you have to set the type to `string?`.

Reply:
1. I was thinking more of a value being absent from the dict, than a special API value.

Reply:
1. Then again, perhaps a better practice would be to model the decoupling that I describe directly in the model: to have an API entity layer, and a model entity layer right below that. Like you might have a Pydantic layer for boundary data validation, but use a more stricly typed object for internal representation.

Reply:
> either it's provided through the API (lsm::all) or it is allocated  
If an attribute is writable by the user, than it is not read-only, and can therefore not be allocated.  cf. a conversation we had in the solutions team regarding this recently: https://inmanta.slack.com/archives/C01QH7140NT/p1710930606322199?thread_ts=1710930381.452039&cid=C01QH7140NT

> Yet you can not express this invariant in the model because you have to set the type to string?
That is true, you have to check yourself in the model if your service was created correctly, which you trust it is if you are in a state that passed validation.

> I was thinking more of a value being absent from the dict, than a special API value.
Which is exactly what `undefined` is in javascript (so in JSON as well), the issue is that you can not represent this in the model, so you can face a situation where the service of the api can not be represented in the model.

Reply:
e.g.

```
entity APIEntity:
    int? allocated_field = null
    lsm::attribute_modifier allocated_field__modifier="r"
end

entity ModelEntity:
    int allocated_field
end


for entity in lsm::all(entity_binding):
    APIEntity(allocated_field=entity["attributes"]["allocated_field"])
end

implementation imp for APIEntity:
    ModelEntity(
        allocated_field=self.allocated_field is defined ? self.allocated_field is defined : allocate(...)
    )
end
```

Major disadvantage: quite some duplication. But perhaps we can find a middle ground somehow. I don't see it yet.

Reply:
> I don't mean writable by the user, I mean provided by the lsm read API, because it was allocated in a previous state.

Okay, I misunderstood you.

> So you end up with static type violations all around

Yep, big downer for the static typing I agree.  But we never need to check everywhere in the model for it, because the value that will be assigned always comes out of the allocator, which will either return a valid allocated value, or `Unknown` if allocation failed previously (unknown value which we only meet in a deallocating state then, where the service is usually not even constructed in the model).

With all of this, I agree that the "nullable" notation of read-only attributes is not ideal, I would prefer if they were not typed like this in the model.  I do still think it would be better to have them typed "nullable" in the api, with a default of `null` than having them sometimes missing from the attributes, but this is a personal opinion.

Coming back to how things are now, and about the remark I made.  I didn't know that the api always added the default `null` value, even if we didn't provided it.  In that case, I see how the `null` default is not required.  To be honest, I am not a fan of it and the "allocation implementation" approach it goes with.  For one thing, it makes it impossible to create the service differently than via the lsm api (one existing use case we have for that is the translation of a service into another, i.e. connect l2Connect and connect CarrierEthernetEvc, two different apis for the same network configuration), and also it doesn't offer the nice separation of DSL domain vs LSM domain (allocation rather belongs in the lather).

Reply:
>  I do still think it would be better to have them typed "nullable" in the api, with a default of null than having them sometimes missing from the attributes, but this is a personal opinion.

I think I agree by now, but I have to consider it some more.

> Coming back to how things are now, and about the remark I made.

You make a good point. Especially combined with the fact that we're both leaning towards explicit defaults in the API in the future, I think it's probably best to stick with explicit model defaults in the model for now, at least in example code, which will establish the best practice.